Source/Global/perform_env.cpp

```diff
@@ -294,8 +307,175 @@ HRESULT CALLBACK HC_PERFORM_ENV::HttpPerformAsyncShimProvider(XAsyncOp op, const
 void CALLBACK HC_PERFORM_ENV::HttpPerformComplete(XAsyncBlock* async)
 {
     HttpPerformContext* performContext{ static_cast<HttpPerformContext*>(async->context) };
-    HRESULT hr = XAsyncGetStatus(async, false);
-    XAsyncComplete(performContext->clientAsyncBlock, hr, 0);
+    XAsyncComplete(performContext->clientAsyncBlock, XAsyncGetStatus(async, false), 0);
+}
+
+struct HC_PERFORM_ENV::WebSocketConnectContext
+{
+    WebSocketConnectContext(
+        HC_PERFORM_ENV* _env,
+        http_internal_string&& _uri,
+        http_internal_string&& _subprotocol,
+        HCWebsocketHandle websocketHandle,
+        XAsyncBlock* _clientAsyncBlock
+    ) : env{ _env },
+        uri{ std::move(_uri) },
+        subprotocol{ std::move(_subprotocol) },
+        websocket{ websocketHandle->websocket },
+        clientAsyncBlock{ _clientAsyncBlock },
+        internalAsyncBlock{ nullptr, this, HC_PERFORM_ENV::WebSocketConnectComplete }
+    {
+    }
+
+    ~WebSocketConnectContext()
+    {
+        if (internalAsyncBlock.queue)
+        {
+            XTaskQueueCloseHandle(internalAsyncBlock.queue);
+        }
+    }
+
+    HC_PERFORM_ENV* const env{};
+    http_internal_string uri;
+    http_internal_string subprotocol;  
+    std::shared_ptr<WebSocket> websocket;
```

Comment:
>std::shared_ptr<WebSocket> [](http://example.com/codeflow?start=4&length=26)

does this need to use mem hook type?

Reply:
the type doesn't need to change, but it is allocated with the mem-hooked allocated in WebSocket::Initialize.